core/src/main/scala/epollcat/internal/ch/SocketHelpers.scala

```diff
@@ -30,8 +30,80 @@ import scala.scalanative.unsafe._
 
 private[ch] object SocketHelpers {
 
-  lazy val preferIPv4Stack =
-    java.lang.Boolean.parseBoolean(System.getProperty("java.net.preferIPv4Stack", "false"))
+  /* An approach with better software engineering becomes available
+   * when epollcat no longer supports Scala Native 0.4.n.
+   * One can eliminate the chance of races & mismatches by allocating a
+   * Scala Native socket and examining its address family. Scala Native
+   * will have done all the work.
+   *
+   * Tcp6Suite.scala uses a similar technique querying an
+   * AsynchronousServerSocketChannel.
+   */
+
+  lazy val useIPv4Stack = {
+    val systemPropertyForcesIPv4 =
+      java.lang.Boolean.parseBoolean(System.getProperty("java.net.preferIPv4Stack", "false"))
+    systemPropertyForcesIPv4 || hasOnlyIPv4Stack()
+  }
+
+  private def hasOnlyIPv4Stack(): Boolean = {
+    val addrinfo = stackalloc[Ptr[posix.netdb.addrinfo]]()
+    val hints = stackalloc[posix.netdb.addrinfo]()
+
+    hints.ai_family = posix.sys.socket.AF_INET6
+    hints.ai_flags = posix.netdb.AI_NUMERICHOST | posix.netdb.AI_NUMERICSERV
+    hints.ai_flags |= posix.netdb.AI_PASSIVE // ServerSocket
+    hints.ai_flags |= posix.netdb.AI_ADDRCONFIG
+    hints.ai_socktype = posix.sys.socket.SOCK_STREAM
+
+    val typelevelOrg6 = c"2606:50c0:8003::153"
+
+    val rtn = posix
+      .netdb
+      .getaddrinfo(
+        typelevelOrg6,
+        c"0",
+        hints,
+        addrinfo
+      )
+
+    val hasIPv6 =
+      try {
+        if (rtn == 0) {
+          // should never happen, but check anyways
+          java.util.Objects.requireNonNull(!addrinfo)
+          (!addrinfo).ai_family == posix.sys.socket.AF_INET6
+        } else {
```

Comment:
Can we assume that the combination of `rtn == 0` and `(!addrinfo).ai_family == posix.sys.socket.AF_INET` is impossible?

If not, is there not the possibility that `addrinfo` contains two entries: the first is `AF_INET` and the second is `AF_INET6`?

Reply:
Thank you for looking at this & keeping me honest.
In the general case you are exactly right.

In this particular case:

Line 53:   `hints.ai_family = posix.sys.socket.AF_INET6`

should cause getaddrinfo to return only AF_INET6 addresses (or else rtn !=0).
Since this is known to be a numeric host 
` val typelevelOrg6 = c"2606:50c0:8003::153` and that is known to be IPv6
it would take perversity for the nameserver to return an IPv4 address.

Are you buying that?

Reply:
I do, thanks! So would it make sense to add a comment that we should never expect this `(!addrinfo).ai_family == posix.sys.socket.AF_INET6` to be `false`?